R/virtualenv.R

```diff
@@ -1,309 +1,260 @@
-      else
-        install_commands <- c(install_commands, "$ sudo apt-get install python-virtualenv")
-    }
-    if (!is.null(install_commands))
-      install_commands <- paste(install_commands, collapse = "\n")
-  } else {
-    if (!have_pip)
-      install_commands <- c(install_commands, "pip")
-    if (!have_virtualenv)
-      install_commands <- c(install_commands, "virtualenv")
-    if (!is.null(install_commands)) {
-      install_commands <- paste("Please install the following Python packages before proceeding:",
-                                paste(install_commands, collapse = ", "))
-    }
-  }
-  if (!is.null(install_commands)) {
-
-    # if these are terminal commands then add special preface
-    if (grepl("^\\$ ", install_commands)) {
-      install_commands <- paste0(
-        "Execute the following at a terminal to install the prerequisites:\n\n",
-        install_commands
-      )
-    }
+  # check for some expected files within virtualenv layout
+  expected <- c("bin/activate", "bin/pip", "bin/python")
+  all(file.exists(file.path(path, expected)))
+}
 
-    stop("Prerequisites for using Python virtualenvs not available.\n\n",
-         install_commands, "\n\n", call. = FALSE)
+
+
+virtualenv_path <- function(envname) {
+
+  # treat environment 'names' containins slashes as paths
+  # rather than environments living in WORKON_HOME
+  if (grepl("[/\\]", envname)) {
+    if (file.exists(envname))
+      envname <- normalizePath(envname, winslash = "/")
+    return(envname)
   }
 
-  # return config
-  list(
-    python = python,
-    virtualenv = virtualenv,
-    pip_version = ifelse(python_version(python) >= "3.0", "pip3", "pip"),
-    root = virtualenv_root()
-  )
+  file.path(virtualenv_root(), envname)
+
 }
 
-python_unix_binary <- function(bin) {
-  locations <- file.path(c("/usr/bin", "/usr/local/bin", path.expand("~/.local/bin")), bin)
-  locations <- locations[file.exists(locations)]
-  if (length(locations) > 0)
-    locations[[1]]
-  else
-    NULL
+
+virtualenv_pip <- function(envname) {
+  file.path(virtualenv_path(envname), "bin/pip")
 }
 
-python_version <- function(python) {
 
-  # check for the version
-  result <- system2(python, "--version", stdout = TRUE, stderr = TRUE)
 
-  # check for error
-  error_status <- attr(result, "status")
-  if (!is.null(error_status))
-    stop("Error ", error_status, " occurred while checking for python version", call. = FALSE)
+virtualenv_default_python <- function(python) {
 
-  # parse out the major and minor version numbers
-  matches <- regexec("^[^ ]+\\s+(\\d+)\\.(\\d+).*$", result)
-  matches <- regmatches(result, matches)[[1]]
-  if (length(matches) != 3)
-    stop("Unable to parse Python version '", result[[1]], "'", call. = FALSE)
+  if (!is.null(python))
+    return(python)
 
-  # return as R numeric version
-  numeric_version(paste(matches[[2]], matches[[3]], sep = "."))
+  config <- py_config()
```

Comment:
We may want to use `py_discover_config()` here. One of the goals of the installation functions is to avoid initializing reticulate in the current process -- this is because if you install into a virtualenv using a default python session then you can't immediately import the Python library (you'd need to restart R). This affects the simplicity of getting started materials (e.g. `install_tensorflow()`, then restart R, then do `library(tensorflow)`.

That's one reason we put in `py_discover_config()` -- it will tell you the version of Python that *would* be loaded without actually loading it.